### PR TITLE
Realtime Publish

### PR DESCRIPTION
**Pull request by tiholic**
Publishing realtime messages either as `name` and `data` or as a `Message` object or `List<Message>`

```dart
//publish name and data
await channel.publish(name: "event1", data: "Yesterday is history, tomorrow is a mystery, but today is a gift");
await channel.publish(name: "event1", data: {"Yesterday": "history", "tomorrow": "mystery", "today": "gift"});
await channel.publish(name: "event1", data: [{"Yesterday": {"is": "history"}, "tomorrow": {"mystery": true}, "day_after": null}, "today", "gift"]);

//publish single message
await channel.publish(message: ably.Message()..name = "event1"..data = {"today": "gift"});

//publish multiple messages
await channel.publish(messages: [
  ably.Message()..name="event1"..data = {"yesterday": "history"},
  ably.Message()..name="event1"..data = {"tomorrow": "mystery"},
  ably.Message()..name="event1"..data = {"today": "gift"}
]);
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -151,6 +151,24 @@ UnSubscribing from channel messages
 await channelMessageSubscription.cancel();
 ```
 
+Publishing channel messages
+
+```dart
+//publish name and data
+await channel.publish(name: "event1", data: "hello world");
+await channel.publish(name: "event1", data: {"hello": "world", "hey": "ably"});
+await channel.publish(name: "event1", data: [{"hello": {"great": "world"}, "ably": {"serious": true, "realtime": true}, "key3": null}, "string4", "string5"]);
```
Comment (zoechi) [inline README.md:160]: This line should be shortened. It requires horizontal scrolling when rendered
https://github.com/ably/ably-flutter/blob/994979557e4288f42412a0ddcc2c714077a01291/README.md#realtime-api

File: example/lib/main.dart [hunk 2]
```diff
@@ -362,6 +370,30 @@ class _MyAppState extends State<MyApp> {
...
   );
 
+  int typeCounter = 0;
+  int realtimePublishCounter = 0;
+  Widget createChannelPublishButton() => FlatButton(
+    onPressed: (_realtimeChannelState==ably.ChannelState.attached)?() async {
+      print('Sendimg rest message...');
+      dynamic data = messagesToPublish[(realtimePublishCounter++ % messagesToPublish.length)];
+      ably.Message m = ably.Message()..data=data..name='Hello';
+      if(typeCounter%3 == 0){
```
Comment (zoechi) [inline example/lib/main.dart:380]: I commented on this in another PR.
Why are these same changes in different PRs?

File: lib/src/impl/realtime/channels.dart [hunk 3]
```diff
@@ -64,12 +64,15 @@ class RealtimePlatformChannel extends PlatformObject implements spec.RealtimeCha
...
-      });
+      Map<String, dynamic> eventPayload = {..._payload};
+      if(messages!=null) {
+        eventPayload['messages'] = messages;
+      } else if(message!=null) {
+        eventPayload['messages'] = [message];
+      }
+      if(name!=null) eventPayload['name'] = name;
+      if(data!=null) eventPayload['data'] = data;
+      await this.invoke(PlatformMethod.publishRealtimeChannelMessage, eventPayload);
```
Comment (zoechi) [inline lib/src/impl/realtime/channels.dart:75]: I'd use collection if 
```dart
    try {
      Map<String, dynamic> eventPayload = {
        ..._payload,
        if (messages != null) 'messages': messages,
        if (messages == null && message != null) 'messages': [message],
        if (name != null) 'name': name,
        if (data != null) 'data': data
      };
      await this
          .invoke(PlatformMethod.publishRealtimeChannelMessage, eventPayload);
    } on PlatformException catch (pe) {
```

File: example/lib/main.dart (hunk 2 above)
Comment (tiholic) [reply example/lib/main.dart:380]: In the next PR this went inside a try-catch as a new issue surfaced.

File: README.md (hunk 1 above)
Comment (tiholic) [reply README.md]: addresseed f0e1f306a2e72900b44d5278a97244300c9372e4

File: lib/src/impl/realtime/channels.dart (hunk 3 above)
Comment (tiholic) [reply lib/src/impl/realtime/channels.dart]: Addressed e6ac991b248dc5412d5873d4e710838a69e52f11

File: android/src/main/java/io/ably/flutter/plugin/AblyMessageCodec.java [hunk 4]
```diff
@@ -136,6 +136,20 @@ private void WriteJsonElement(ByteArrayOutputStream stream, JsonElement value){
         }
     }
 
+    /**
+     * Converts Map to JsonObject, ArrayList to JsonArray and
+     * returns null if these 2 types are a no-match
+     * */
+    static JsonElement readValueAsJsonElement(final Object object) {
+        Gson gson = new Gson();
```
Comment (QuintinWillison) [inline android/src/main/java/io/ably/flutter/plugin/AblyMessageCodec.java:144]: I recognise the scope is short here but I would still prefer this reference were `final`.

Comment (tiholic) [reply android/src/main/java/io/ably/flutter/plugin/AblyMessageCodec.java]: Sure, thanks for pointing this out. I'm making variables final whichever I am coming across and feel is appropriate 😄 

Addressed 59ff55b9cce0473920b322e21f6b9297a3954574

File: README.md [hunk 5]
```diff
@@ -151,6 +151,24 @@ UnSubscribing from channel messages
 await channelMessageSubscription.cancel();
 ```
 
+Publishing channel messages
+
+```dart
+//publish name and data
```
Comment (QuintinWillison) [inline README.md:157]: Please always include a space after `//` comment openers. e.g., in this case:

```suggestion
// publish name and data
```

File: android/src/main/java/io/ably/flutter/plugin/AblyFlutterMessage.java [hunk 6]
```diff
@@ -11,6 +13,13 @@
         this.handle = handle;
         this.message = message;
     }
+
+    @NonNull
+    @Override
+    public String toString() {
+        return "{message=" + (message.toString()) + ", handle=" + (handle==null?"":handle.toString()) + "}";
```
Comment (QuintinWillison) [inline android/src/main/java/io/ably/flutter/plugin/AblyFlutterMessage.java:20]: Please include spaces for clarity. This is too compressed. This is better:

```suggestion
        return "{message=" + (message.toString()) + ", handle=" + (handle == null ? "" : handle.toString()) + "}";
```

File: android/src/main/java/io/ably/flutter/plugin/AblyMessageCodec.java [hunk 7]
```diff
@@ -136,6 +136,20 @@ private void WriteJsonElement(ByteArrayOutputStream stream, JsonElement value){
         }
     }
 
+    /**
+     * Converts Map to JsonObject, ArrayList to JsonArray and
+     * returns null if these 2 types are a no-match
+     * */
```
Comment (QuintinWillison) [inline android/src/main/java/io/ably/flutter/plugin/AblyMessageCodec.java:142]: What's with the space in the middle here? I would have expected:

```suggestion
     **/
```

File: android/src/main/java/io/ably/flutter/plugin/AblyMessageCodec.java [hunk 8]
```diff
@@ -136,6 +136,20 @@ private void WriteJsonElement(ByteArrayOutputStream stream, JsonElement value){
...
+    static JsonElement readValueAsJsonElement(final Object object) {
+        final Gson gson = new Gson();
+        if(object instanceof Map) {
```
Comment (QuintinWillison) [inline android/src/main/java/io/ably/flutter/plugin/AblyMessageCodec.java:145]: Please leave a space after `if` statements - i.e. in this case:

```suggestion
        if (object instanceof Map) {
```

File: android/src/main/java/io/ably/flutter/plugin/AblyMessageCodec.java [hunk 9]
```diff
@@ -136,6 +136,20 @@ private void WriteJsonElement(ByteArrayOutputStream stream, JsonElement value){
...
+            return gson.fromJson(gson.toJson(object, Map.class), JsonObject.class);
+        } else if (object instanceof ArrayList){
```
Comment (QuintinWillison) [inline android/src/main/java/io/ably/flutter/plugin/AblyMessageCodec.java:147]: Please include a space before opening curly braces. i.e.:

```suggestion
        } else if (object instanceof ArrayList) {
```

File: android/src/main/java/io/ably/flutter/plugin/AblyMessageCodec.java [hunk 10]
```diff
@@ -241,6 +255,24 @@ private TokenDetails readTokenDetails(Map<String, Object> jsonMap) {
         return o;
     }
 
+    private Message decodeChannelMessage(Map<String, Object> jsonMap){
+        if(jsonMap==null) return null;
```
Comment (QuintinWillison) [inline android/src/main/java/io/ably/flutter/plugin/AblyMessageCodec.java:259]: Again, spacing after `if`, please. i.e.:

```suggestion
        if (jsonMap==null) return null;
```

File: android/src/main/java/io/ably/flutter/plugin/AblyMethodCallHandler.java [hunk 11]
```diff
@@ -256,6 +262,48 @@ private void setRealtimeChannelOptions(
         });
     }
 
+    private void publishRealtimeChannelMessage(
```
[the remark this answers is not in the document]
Comment (tiholic) [reply android/src/main/java/io/ably/flutter/plugin/AblyMethodCallHandler.java:265]: @QuintinWillison I've made necessary variables final as applicable in all the Java files

a2540e49f0ba1cffed90d70bdd14ca242106d514

File: example/test_harness/listen.js [hunk 12]
```diff
@@ -0,0 +1,15 @@
+const Ably = require('ably');
+
+if(process.argv[2] && !process.argv[3]) throw new Error("Invalid number of args: pass either 0 or 2 args");
+
+const key = process.argv[2];
+const channelName = process.argv[3] || "test-channel";
+const environment = 'sandbox';
+const logLevel = 2;
+
+var realtime = new Ably.Realtime({key, environment, logLevel});
```
[the remark this answers is not in the document]
Comment (tiholic) [reply example/test_harness/listen.js]: test-harness code improvised:
1. var -> const
2. prettier: using [prettier playground](https://prettier.io/playground/) - as we could have very rare edits to these files.
_Default rules applied_

05040250d43423b4497223a7b9f719cc7a44df58

File: example/test_harness/listen.js [hunk 13]
```diff
@@ -0,0 +1,15 @@
+const Ably = require('ably');
+
+if(process.argv[2] && !process.argv[3]) throw new Error("Invalid number of args: pass either 0 or 2 args");
```
[the remark this answers is not in the document]
Comment (tiholic) [reply example/test_harness/listen.js]: Resolved as in this comment: https://github.com/ably/ably-flutter/pull/23#discussion_r471909937

0504025

File: README.md (hunk 5 above)
Comment (tiholic) [reply README.md]: Applied this suggestion on all comments

e8eab0f6e60e99c4e8811943ab2b7a801e064ce4

File: android/src/main/java/io/ably/flutter/plugin/AblyMessageCodec.java (hunk 7 above)
Comment (tiholic) [reply android/src/main/java/io/ably/flutter/plugin/AblyMessageCodec.java]: Addressed all java code formatting

052467d338e6fd8a37e3474c03b355e0ec1b8fdc

File: android/src/main/java/io/ably/flutter/plugin/AblyMessageCodec.java (hunk 8 above)
Comment (tiholic) [reply android/src/main/java/io/ably/flutter/plugin/AblyMessageCodec.java]: Same as https://github.com/ably/ably-flutter/pull/23#discussion_r471921091

052467d338e6fd8a37e3474c03b355e0ec1b8fdc

File: android/src/main/java/io/ably/flutter/plugin/AblyMessageCodec.java (hunk 9 above)
Comment (tiholic) [reply android/src/main/java/io/ably/flutter/plugin/AblyMessageCodec.java]: Same as https://github.com/ably/ably-flutter/pull/23#discussion_r471921091

052467d338e6fd8a37e3474c03b355e0ec1b8fdc

File: android/src/main/java/io/ably/flutter/plugin/AblyMessageCodec.java (hunk 10 above)
Comment (tiholic) [reply android/src/main/java/io/ably/flutter/plugin/AblyMessageCodec.java]: Same as https://github.com/ably/ably-flutter/pull/23#discussion_r471921091

052467d338e6fd8a37e3474c03b355e0ec1b8fdc

File: android/src/main/java/io/ably/flutter/plugin/AblyFlutterMessage.java (hunk 6 above)
Comment (tiholic) [reply android/src/main/java/io/ably/flutter/plugin/AblyFlutterMessage.java]: Same as https://github.com/ably/ably-flutter/pull/23#discussion_r471921091

052467d338e6fd8a37e3474c03b355e0ec1b8fdc

File: ios/Classes/AblyFlutterPlugin.m [hunk 14]
```diff
@@ -143,6 +143,40 @@ -(void)registerWithCompletionHandler:(FlutterResult)completionHandler;
...
+    ARTRealtime *const realtimeWithHandle = [ably realtimeWithHandle: realtimeHandle];
+
+    NSDictionary *const realtimePayload = data.message;
+    NSString *const channelName = (NSString*)[realtimePayload objectForKey:@"channel"];
+    ARTChannelOptions *const channelOptions = (ARTChannelOptions*)[realtimePayload objectForKey:@"options"];
+    ARTRealtimeChannel *const channel = [realtimeWithHandle.channels get:channelName options:channelOptions];
+    void (^callback)(ARTErrorInfo *_Nullable) = ^(ARTErrorInfo *_Nullable error){
+        if(error){
+            result([
+                    FlutterError
```
[the remark this answers is not in the document]
Comment (tiholic) [reply ios/Classes/AblyFlutterPlugin.m]: Addressed f265e079c64888c8d323ee7bc719e07ff47625bd

[the remark this answers is not in the document]
Comment (tiholic) [reply ios/Classes/AblyFlutterPlugin.m]: Applied re-indent code on all iOS related files eff54ca16c0ec9e63db5047781dd445cd353dd77

**Pull request merged**